### ISSUE TITLE
[native-image] [tracer] Potential bug in the merge function of the tracer

### ISSUE THREAD

**Issue by neomatrix369** [oracle/graal#1599]
I think there is a likelihood that the tracer somehow overwrites the manually amended sections of the config file.

Below are two scenarios, the first one, is a result of a manual amendment after a runtime recommendation that the two fields need write permissions, so I did that. Due to another reason, re-ran the tracer and it resulted in the amended block been replaced by the original block.

Before running trace (after manual amendment):

_reflect-config.json_
```json
{
  "name":"org.neomatrix369.common.Library",
  "allDeclaredFields":true,
  "fields":[
    {"name":"name", "allowWrite" : true},
	{"name":"field", "allowWrite" : true}
  ]
}
```

After running trace again:

_reflect-config.json_
```json
{
  "name":"org.neomatrix369.common.Library",
  "allDeclaredFields":true
}
```

And this is fairly reproducible, I have already done this a couple of times and I see that it consistently replaces it with the original block.

Let me know if you need anything else to track this issue down.

**Comment by peter-hofer** [comment]
Thanks for your report @neomatrix369! It's possible that the merge code assumes that the two fields are covered by `allDeclaredFields` without considering that this discards their `allowWrite` attribute. Can you provide command lines and source code or a JAR file to reproduce (ideally a minimal example)?

**Comment by neomatrix369** [comment]
> Thanks for your report @neomatrix369! It's possible that the merge code assumes that the two fields are covered by `allDeclaredFields` without considering that this discards their `allowWrite` attribute. Can you provide command lines and source code or a JAR file to reproduce (ideally a minimal example)?

I'll try to provide some example. Quite possible and shouldn't happen since `allDeclaredFields` and `allowWrite` are two different levels of abstractions if I understand right, one if read-only and the other one allows writing.

**Comment by neomatrix369** [comment]
@peter-hofer Sorry for the delay I have an example project for you where you can reproduce the reported bug, I'll upload it asap.

**Comment by peter-hofer** [comment]
Thank you @neomatrix369 , I'll look into it!

**Comment by neomatrix369** [comment]
> Thank you @neomatrix369 , I'll look into it!

Should work out of the box - otherwise please let me know.

**Comment by peter-hofer** [comment]
Fixed in 429c3590c85ba8064cefc66badeba412560554e7 - thanks for your concise example!

**Comment by neomatrix369** [comment]
> Fixed in [429c359](https://github.com/oracle/graal/commit/429c3590c85ba8064cefc66badeba412560554e7) - thanks for your concise example!

Glad it was of help. Good work, quite a quick turnaround! 👏 

I'm thinking some kind of test in the test suite put in place to check for regression, just in case.
